### DISCUSSION TITLE
Shutting down the uvicorn server master from a FastAPI worker

### DISCUSSION THREAD

**Discussion by ocanty** [fastapi/fastapi#7458] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

I want to kill the entire uvicorn server whenever one of my workers encounters an issue it cannot resolve.

I currently run fastapi inside the bundled Docker container recommended in the documentation, when my worker cannot connect to an upstream api because it is down/unavailable, I can call sys.quit(-1) to kill the worker with an error code, but this only causes uvicorn to restart the worker.

Is there any interface to kill the uvicorn master server and return an error code so that it kills my Docker container? The only alternative I can see is using health checks instead of killing the container

![image](https://user-images.githubusercontent.com/14853770/83693821-b7a40280-a5ee-11ea-9407-b998b8e506bf.png)

One solution I have found would be using os.kill to kill pid 1/sending a SIGTERM but I'm not sure if ASGI may have a procedure/interface for this rather than bluntly signalling the process. I have searched the uvicorn and Starlette docs and found nothing

**Comment by tiangolo** [comment]
I imagine you solved your problem, so thanks for closing the issue :+1:

**Comment by nicoinn** [comment]
@ocanty  Could you consider sharing your solution ? I have exactly the same question...

**Comment by cormacp** [comment]
Likewise, I've the same question, any insight into the best way to do this programatically?

**Comment by ChoOo7** [comment]
I found a very dirty way : 

def crash():
    try:
        crash()
    except:
        crash()

My app is inside a docker container with auto restart
so.... a fresh container start after

**Comment by janheindejong** [comment]
Same... what's the solution here? :-)

**Comment by JaDogg** [comment, marked as answer]
How about this?

```python
api = FastAPI()


def self_terminate():
    time.sleep(1)
    parent = psutil.Process(psutil.Process(os.getpid()).ppid())
    parent.kill()


@api.post("/kill")
async def kill():
    threading.Thread(target=self_terminate, daemon=True).start()
    return {"success": True}
```

I'm using psutil to get parent process and kill it. 👍🏽 This sends a signal to parent process. This worked for me in Windows. (Should work in Linux/MacOS but not sure) You can also send parent process a SIGINT to simulate Ctrl+C 😎
